File: examples/java-examples/src/main/java/com/edgedb/examples/AbstractTypes.java
package com.edgedb.examples;

import com.edgedb.driver.GelClientPool;
import com.edgedb.driver.annotations.GelType;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.concurrent.CompletionStage;

public class AbstractTypes implements Example {
    private static final Logger logger = LoggerFactory.getLogger(AbstractTypes.class);

    @GelType
    public static abstract class Media {
        public String title;
    }

    @GelType
    public static class Show extends Media {
        public Long seasons;
    }

    @GelType
    public static class Movie extends Media {
        public Long releaseYear;
    }

    @Override
    public CompletionStage<Void> run(GelClientPool clientPool) {
        return clientPool
                .execute("insert Movie { title := \"The Matrix\", release_year := 1999 } unless conflict on .title")
                .thenCompose(v -> clientPool.execute("insert Show { title := \"The Office\", seasons := 9 } unless conflict on .title"))
                .thenCompose(v -> clientPool.query(Media.class, "select Media { title, [is Movie].release_year, [is Show].seasons }"))
                .thenAccept(content -> {
                    for (var media : content) {
                        if(media instanceof Show) {
                            logger.info("Got show {}", media);
                        }

                        if(media instanceof Movie) {
                            logger.info("Got movie {}", media);
                        }
                    }
                });
    }
}

File: examples/kotlin-examples/src/main/kotlin/com/edgedb/examples/AbstractTypes.kt
package com.edgedb.examples

import com.edgedb.driver.GelClientPool
import com.edgedb.driver.annotations.GelType
import kotlinx.coroutines.future.await
import org.slf4j.LoggerFactory

class AbstractTypes : Example {
    companion object {
        private val logger = LoggerFactory.getLogger(AbstractTypes::class.java)!!
    }

    @GelType
    abstract class Media {
        var title: String? = null
    }

    @GelType
    class Show : Media() {
        var seasons: Long? = null
    }

    @GelType
    class Movie : Media() {
        var releaseYear: Long? = null
    }

    override suspend fun runAsync(clientPool: GelClientPool) {
        clientPool.execute(
                """insert Movie { 
                   title := "The Matrix", 
                   release_year := 1999 
               } unless conflict on .title;
               insert Show { 
                   title := "The Office", 
                   seasons := 9 
               } unless conflict on .title"""
        ).await()

        val results = clientPool.query(
                Media::class.java,
                """select Media { 
                   title, 
                   [is Movie].release_year, 
                   [is Show].seasons 
               }"""
        ).await()

        for(result in results) {
            if(result is Show) {
                logger.info(
                        "Got show: title: {}, seasons: {}",
                        result.title, result.seasons
                )
            } else if(result is Movie) {
                logger.info(
                        "Got movie: title: {}, release year: {}",
                        result.title, result.releaseYear
                )
            }
        }
    }
}